src/hazelweb/layout/UHDoc_common.re

```diff
@@ -140,99 +141,139 @@ type formatted_child =
   | EnforcedInline(t)
   | Unformatted((~enforce_inline: bool) => t);
 
-let pad_bidelimited_open_child =
-    (~inline_padding: (t, t)=(empty_, empty_), child: formatted_child): t => {
-  open Doc;
-  let inline_choice = child_doc => {
-    let (left, right) = inline_padding;
-    let lpadding = left == empty_ ? [] : [left];
-    let rpadding = right == empty_ ? [] : [right];
-    hcats([
-      hcats(List.concat([lpadding, [child_doc], rpadding]))
-      |> annot(UHAnnot.OpenChild(InlineWithBorder)),
-    ]);
-  };
-  let para_choice = child_doc =>
-    child_doc |> indent_and_align |> annot(UHAnnot.OpenChild(Multiline));
-  switch (child) {
-  | EnforcedInline(child_doc) => inline_choice(child_doc)
-  | UserNewline(child_doc) =>
-    hcats([user_newline, linebreak(), para_choice(child_doc), linebreak()])
-  | Unformatted(formattable_child) =>
-    choices([
-      inline_choice(formattable_child(~enforce_inline=true)),
-      hcats([
-        linebreak(),
-        para_choice(formattable_child(~enforce_inline=false)),
-        linebreak(),
-      ]),
-    ])
-  };
+let pad = ((left, right), ~newline=[], child_doc) => {
+  let lpadding = left == empty_ ? [] : [left];
+  let rpadding = right == empty_ ? [] : [right];
+  Doc.hcats(newline @ lpadding @ [child_doc] @ rpadding);
 };
 
-let pad_closed_child =
+let pad_open_inline_child = (inline_padding, with_border, child_doc) => {
```

Comment:
The preexisting padding functions had some awkwardly duplicated logic, and did not have full combinatorial coverage, including the left-delimited closed-child case, which is necessary here for type annotations. The refactoring was necessary to prevent more copy-paste code here.